presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileReader.java

```diff
@@ -0,0 +1,584 @@
+        }
+        input.readBytes(compressedHeaderBuffer, 0, compressedHeaderSize);
+
+        // decompress block header
+        Slice header;
+        if (codec != null) {
+            if (headerBuffer.length() < uncompressedHeaderSize) {
+                headerBuffer = Slices.allocate(uncompressedHeaderSize);
+            }
+            Slice buffer = headerBuffer.slice(0, uncompressedHeaderSize);
+
+            codec.decompress(compressedHeaderBuffer, buffer);
+
+            header = buffer;
+        }
+        else {
+            header = compressedHeaderBuffer.slice(0, uncompressedHeaderSize);
+        }
+        BasicSliceInput headerInput = header.getInput();
+
+        // read number of rows in block
+        blockRowCount = Ints.checkedCast(readVInt(headerInput));
+        rowsRead += blockRowCount;
+        blockPosition = 0;
+        currentChunkSize = min(ColumnData.MAX_SIZE, blockRowCount);
+
+        // set column buffers
+        for (int columnIndex = 0; columnIndex < columnCount; columnIndex++) {
+            int compressedDataSize = Ints.checkedCast(readVInt(headerInput));
+            int uncompressedDataSize = Ints.checkedCast(readVInt(headerInput));
+            if (codec == null && compressedDataSize != uncompressedDataSize) {
+                throw corrupt("Invalid RCFile %s", dataSource);
+            }
+
+            int lengthsSize = Ints.checkedCast(readVInt(headerInput));
+
+            Slice lengthsBuffer = headerInput.readSlice(lengthsSize);
+
+            Slice dataBuffer;
+            if (readColumns.contains(columnIndex)) {
+                dataBuffer = input.readSlice(compressedDataSize);
+            }
+            else {
+                input.skip(compressedDataSize);
+                dataBuffer = Slices.EMPTY_SLICE;
+            }
+
+            columns[columnIndex].setBuffers(lengthsBuffer, dataBuffer, uncompressedDataSize);
+        }
+
+        return currentChunkSize;
+    }
+
+    public ColumnData readColumnData(int columnIndex)
+            throws IOException
+    {
+        if (currentChunkSize <= 0) {
+            throw new IllegalStateException("No more data");
+        }
+        return columns[columnIndex].readColumnData(blockPosition, currentChunkSize);
+    }
+
+    private void seekToFirstBlockIn(long offset, long length)
+            throws IOException
+    {
+        if (length < SYNC_SEQUENCE_LENGTH) {
+            closeQuietly();
+            return;
+        }
+
+        Slice sync = getSync();
+
+        // read 1 MB chunks at a time, but only skip ahead 1 MB - MIN_SYNC_SEQUENCE_LENGTH bytes
+        // this causes a re-read of MIN_SYNC_SEQUENCE_LENGTH bytes each time, but is much simpler code
+        byte[] buffer = new byte[Ints.checkedCast(Math.max(1 << 20, length))];
+        Slice bufferSlice = Slices.wrappedBuffer(buffer);
+        for (long position = 0; position < length - SYNC_SEQUENCE_LENGTH; position += bufferSlice.length() - SYNC_SEQUENCE_LENGTH) {
```

Comment:
It seems like this code does not work with sync+len that's on the boundary of a split. Neither split will read such a row group.